rust/common-wasi/src/stdio/worker_thread_stdin.rs

```diff
@@ -0,0 +1,177 @@
+//! Handling for standard in using a worker task.
+//!
+//! Standard input is a global singleton resource for the entire program which
+//! needs special care. Currently this implementation adheres to a few
+//! constraints which make this nontrivial to implement.
+//!
+//! * Any number of guest wasm programs can read stdin. While this doesn't make
+//!   a ton of sense semantically they shouldn't block forever. Instead it's a
+//!   race to see who actually reads which parts of stdin.
+//!
+//! * Data from stdin isn't actually read unless requested. This is done to try
+//!   to be a good neighbor to others running in the process. Under the
+//!   assumption that most programs have one "thing" which reads stdin the
+//!   actual consumption of bytes is delayed until the wasm guest is dynamically
+//!   chosen to be that "thing". Before that data from stdin is not consumed to
+//!   avoid taking it from other components in the process.
+//!
+//! * Tokio's documentation indicates that "interactive stdin" is best done with
+//!   a helper thread to avoid blocking shutdown of the event loop. That's
+//!   respected here where all stdin reading happens on a blocking helper thread
+//!   that, at this time, is never shut down.
+//!
+//! This module is one that's likely to change over time though as new systems
+//! are encountered along with preexisting bugs.
+
+use crate::poll::Subscribe;
+use crate::stdio::StdinStream;
+use crate::{HostInputStream, StreamError};
+use bytes::{Bytes, BytesMut};
+use std::io::{IsTerminal, Read};
+use std::mem;
+use std::sync::{Condvar, Mutex, OnceLock};
+use tokio::sync::Notify;
```

Comment:
This uses host stdin, and should only ever operate on the input stream (if any) provided to WasiCtx, as inheriting streams has been removed.